heat/core/operations.py

```diff
@@ -510,7 +533,7 @@ def __reduce_op(x, partial_op, reduction_op, axis, out):
     if x.split is not None and (axis is None or axis == x.split):
         split = None
         if x.comm.is_distributed():
-            x.comm.Allreduce(MPI.IN_PLACE, partial[0], reduction_op)
+            x.comm.Allreduce(MPI.IN_PLACE, partial, reduction_op)
```

Comment:
This doesn't seem to work when distributed and reduction_op is MPI_ARGMIN. Try this test with mpirun -n 2:
```
data = ht.array(ht.random.randn(4, 5), split=0)
result = ht.argmin(data, axis=0)
print('rank {} ht.argmin final result: {}'.format(data.comm.rank, result))
```
Do not set the seed!
The result from Allreduce will only contain min indices from the second half of the tensor (rank 1):

```
tensor([[-1.5692,  0.9294, -0.0859,  0.8452, -1.9167],
        [-0.9213,  0.8887,  0.8188,  1.1889,  0.1289],
        [ 0.5067, -0.6628,  0.7653, -0.6449,  2.5042],
        [ 0.6080, -1.7988,  0.5752,  0.2830,  1.1340]])
tensor([[ 1.7337, -2.8722,  0.5502,  1.9691, -0.3482],
        [ 0.2360,  1.5681, -1.3815,  2.6588, -0.7795],
        [ 0.7685,  0.7748, -1.2191, -0.1385, -0.6562],
        [-0.4852, -0.4101, -0.6436,  0.0545, -0.2059]])
rank 0 ht.argmin final result: tensor([[7, 4, 5, 6, 5]])
rank 1 ht.argmin final result: tensor([[7, 4, 5, 6, 5]])

```
Note that the actual MPI_ARGMIN result is correct:
```
MPI_ARGMIN: rhs.storage:   
-1.5691732168197632
 -2.872173547744751
 -1.3814921379089355
 -0.6449006795883179
 -1.9167150259017944
 0.0
 4.0
 5.0
 2.0
 0.0
```

The tests didn't catch it because  we've set torch.manual_seed(1), so the tensors on each rank are identical, and the result (all min indices from highest rank) isn't that suspicious.

Reply:
Is this related to the problem with randn? (#110) since currently randn will simply create a tensor on each node that is the same size (the given size). 

what I'm concerned about is this: the tensors generated by randn are not created correctly.

Reply:
Hi Daniel,

but in this case the random tensor is "wrapped" in a ht.array with split=0 and that behaves as expected, as far as I can tell.

Reply:
I will have a look into it.

Reply:
In fact, the MPI_ARGMIN implementation seems to be incorrect. It has nothing to do with randn() in any way. Consider this:

`
# foo.py
import heat as ht

data = ht.array(ht.random.randn(4, 5), split=0)
print(data, data.shape, data.lshape)
print(ht.argmin(data, axis=0))
`

And:

`
mpirun --tag-output -np 2 python foo.py

[1,0]<stdout>:tensor([[ 0.3886,  0.7388,  1.7783,  0.6372, -1.3354],
[1,0]<stdout>:        [-0.3796,  1.6146, -1.5172,  1.7657,  0.0242],
[1,0]<stdout>:        [ 0.1005, -0.8079,  0.4624,  0.8972, -0.5272],
[1,0]<stdout>:        [ 0.2935,  0.3631,  2.3646,  1.1830, -0.0578]]) (8, 5) (4, 5)
[1,1]<stdout>:tensor([[-1.4479, -0.6387,  0.4102,  1.0344, -0.0191],
[1,1]<stdout>:        [-0.4462, -1.2580,  1.4736,  0.3634,  0.2449],
[1,1]<stdout>:        [-0.3439,  1.1679, -1.0671, -1.2921,  1.1171],
[1,1]<stdout>:        [-1.7884,  0.1019, -1.3534, -0.2997,  0.4900]]) (8, 5) (4, 5)
[1,0]<stdout>:tensor([[-1.7884, -1.2580, -1.3534, -1.2921, -0.0191],
[1,0]<stdout>:        [ 7.0000,  5.0000,  7.0000,  6.0000,  4.0000]], dtype=torch.float64)
[1,1]<stdout>:tensor([[-1.7884, -1.2580, -1.3534, -1.2921, -0.0191],
[1,1]<stdout>:        [ 7.0000,  5.0000,  7.0000,  6.0000,  4.0000]], dtype=torch.float64)
[1,0]<stdout>:tensor([[7, 5, 7, 6, 4]])
[1,1]<stdout>:tensor([[7, 5, 7, 6, 4]])
`

In the lines 9-12 you the output of the reduction result from MPI_ARGMIN. For the third column (index=2) you already see that the reduction result is wrong. -1.3534 is obviously not smaller than -1.5172.

Reply:
Hi Markus, glad you can reproduce it.

I find that rhs.storage within MPI_ARGMIN contains the correct result (see above). So I'm not sure if it's a problem with MPI_ARGMIN or with the Allreduce call.

Reply:
The issue is the write back into the buffer

`
rhs.storage().copy_(result.storage())
`

This needs to be fixed

Reply:
Provided fix, to be found in MPI_ARGMIN